src/expr/src/scalar/mod.rs

```diff
@@ -1961,6 +1982,129 @@ impl VisitChildren<Self> for MirScalarExpr {
     }
 }
 
+/// Filter characteristics that are used for ordering join inputs.
+/// This can be created for a `Vec<MirScalarExpr>`, which represents an AND of predicates.
+///
+/// The fields should be ordered by typical selectivity, so that Ord gives the right
+/// ordering for join inputs. Bigger is better, i.e., will tend to come to the front of the inputs.
+#[derive(Eq, PartialEq, Ord, PartialOrd, Debug, Clone, Serialize, Deserialize, Hash, MzReflect)]
+pub struct FilterCharacteristics {
+    literal_equality: bool,
```

Comment:
`literal_inequality` is `usize` to handle `BETWEEN`.

For `literal_equality`, you are right that `#0 = 0 AND #1 = 2` could benefit from `literal_equality` also being `usize`, but the issue is that literal equalities often appear in ORs, e.g. an IN list is translated to `#0 = 5 OR #0 = 7`, in which case more of them means less selectivity. Of course, this can also happen with literal inequalities (that they appear in an OR), but it's maybe less typical for those. These things are very heuristic, sorry for that! We could tweak these characteristics endlessly, but I think the long-term solution will be to measure filter selectivity at run time (or at least gather statistics by sampling stuff that is in persist).